hilti/toolchain/src/compiler/coercion.cc

```diff
@@ -424,6 +424,22 @@ struct VisitorType : public visitor::PreOrder<std::optional<Type>, VisitorType>
     }
 
     result_t operator()(const type::Optional& r) {
+        if ( auto t = dst.tryAs<type::Optional>() ) {
+            const auto& s = r.dereferencedType();
+            const auto& d = t->dereferencedType();
+
+            if ( type::sameExceptForConstness(s, d) ) {
+                if ( style & CoercionStyle::Assignment )
+                    // Assignments copy, so it's safe to turn even into the
+                    // destination without considering constness.
+                    return dst;
+
+                if ( type::isConstant(s) && type::isConstant(d) )
+                    // It's safe to turn a non-const value into a const value.
```

Comment:
I am confused by the comment as the condition seems to be _safe to turn const value into const value_. If I change the code to match the comment `if ( ! type::isConstant(s) && type::isConstant(d) )` the tests still pass, so it would be great if we had something in the test suite which checks non-assignment behavior if possible.

Reply:
Yeah, that's a typo. That said, turns out I can't come up with an example where this is actually needed. So I'll just take this out until we find a case where's it's needed.